### PR TITLE
Allow Unordered Dimensions for DimsSchema Validation

### PR DESCRIPTION
**Pull request by raymondwjang**
Ref: #5 

## Problem Statement
Xarray supports named dimensions; this is one of the main features of Xarray and it allows users to keep track of dimensions by their names rather than their positions:
```python
arr = np.ones((5, 5))
da = xr.DataArray(arr, dims=["dim0", "dim1"])
da.sel(dim0=3)  # xarray
# instead of
arr[0][3]  # numpy
```
This holds true in Xarray's internal functionalities, such as dot / cross products, where the operations occur along the dimensions of the same name, rather than the lining up the last dimension of the former array to the first dimension of the latter array. In some use cases, this effectively eliminates the need for transposing Xarray DataArrays, i.e. arranging dimensions in a correct order.

As such, it would make sense that in some cases of Xarray usage, users will care less about the order of dimensions, and rather focus on matching the set of the names.

Implementing this without directly manipulating xarray-validate is difficult, since it requires the user to test against all permutations of the dimensions and checking if any of them passes.

## Proposal

The feature proposed here works like below:
```python
arr1 = xr.DataArray(np.ones((5, 5)), dims=('dim0', 'dim1'))
dims_schema = DimsSchema(('dims1', 'dims0'), ordered=False)

dims_schema.validate(arr1)  # returns None

arr1 = xr.DataArray(np.ones((5, 5)), dims=('dim0', 'dim1'))
dims_schema = DimsSchema(('dims1', 'dims0'), ordered=True)  # ordered defaults to True, and thus can be omitted.

dims_schema.validate(arr1)  # SchemaError
```

## Implementation

The implementation details are rather simple:
```python
@_attrs.define(on_setattr=[_attrs.setters.convert, _attrs.setters.validate])
class DimsSchema(BaseSchema):
    ...

# ------- ADDITION ------- #

    ordered: bool = _attrs.field(default=True)

# --------- END ---------- #
    ...

    def validate(self, dims: DimsT) -> None:
        # Inherit docstring

        if len(self.dims) != len(dims):
            raise SchemaError(
                f"dimension number mismatch: got {len(dims)}, expected {len(self.dims)}"
            )

# ------- ADDITION ------- #

        if not self.ordered:
            actual = set(dims)
            expected = set(self.dims)

            if actual != expected:
                raise SchemaError(
                    f"dimension mismatch: expected but not received {expected - actual}, and not expected but received {actual - expected}"
                )

        else:

# --------- END ---------- #

            for i, (actual, expected) in enumerate(zip(dims, self.dims)):
                if expected is not None and actual != expected:
                    raise SchemaError(
                        f"dimension mismatch in axis {i}: got {actual}, expected {expected}"
                    )
```

## Tests

I've added two unit tests, which focuses on `ordered=False` cases. The existing tests need deeper changes to allow negative tests (where the validation fails), so I have not added those yet.

### CODE REVIEW

File: tests/test_components.py [hunk 1]
```diff
@@ -98,7 +110,10 @@ def test_component_schema(component, schema_args, validate, json):
     """
     # Initialization
     try:
-        schema = component(schema_args)
+        if isinstance(schema_args, dict) and component in [DimsSchema]:
```
Comment (raymondwjang) [inline tests/test_components.py:113]: not a fan of this special case but couldn't figure out a better way 🥀

Comment (leroyvn) [reply tests/test_components.py:113]: This is because by adding a kwarg to the `DimsSchema` constructor, you created an edge case in the initially foreseen init / (de)serialization pattern. To deal with this properly, best is IMO to:

* update the (de)serialization protocol of `DimsSchema` to also support the case `ordered=False`;
* give `DimsSchema` its own testing functions with a clean and simple initialization sequence.

I gave this a try, I think it works. I will open a PR that supersedes this one.

File: src/xarray_validate/components.py [hunk 2]
```diff
@@ -92,12 +93,25 @@ def validate(self, dims: DimsT) -> None:
...
+            actual = set(dims)
+            expected = set(self.dims)
+
+            if actual != expected:
                 raise SchemaError(
-                    f"dimension mismatch in axis {i}: got {actual}, expected {expected}"
+                    "dimension mismatch: expected but not received "
+                    f"{expected - actual}, "
+                    f"and not expected but received {actual - expected}"
                 )
```
Comment (leroyvn) [inline src/xarray_validate/components.py:105]: That won't work with wildcards: the validator will expect a `None` dimension, which will never come. I think we need similar logic to the ordered case to handle that.

**Review by leroyvn — COMMENTED** [review]
Hi @raymondwjang, thanks a lot for your input. There are a few corner cases that require some deeper changes. I took care of that, you can have a look in the PR I will open.

**Comment by leroyvn** [comment]
See #10
